testbed/tests/brakelight_test.py

```diff
@@ -0,0 +1,46 @@
+# firmware/testbed/tests/brakelight_test.py
+
+import time
+from testbeds.hil_testbed import teststand
+from drivers.common_drivers.can_driver import HILBoard
+from tqdm import tqdm
+
+# Define constants for the threshold voltage when the brake light turns on
+BRAKE_LIGHT_ON_THRESHOLD_VOLTAGE_1 = 833 # Lower threshold voltage
+BRAKE_LIGHT_ON_THRESHOLD_VOLTAGE_2 = 841 # Upper threshold voltage
+
+def test_brakelight(teststand):
+    print(teststand.vehicle_boards)
+    
+    vcu_hil: HILBoard = teststand.hil_boards["vcu_hil"]
+    # update pud -> pdu_hil
+    pdu_hil: HILBoard = teststand.hil_boards["pdu_hil"]
+
+    # Iterate over the entire voltage range
+    for voltage in tqdm(range(0, 3301)):
+    
+        # Set the brake position voltage
+        vcu_hil.set_signal("BrakePosition", {"BrakePosition": voltage}) # Reference to HIL.dbc line 69
+        time.sleep(0.001)
+        
+        # Check the Brake Pos Status
+        assert vcu_hil.get_signal("BrakePosStatus"), " the brake position status signal is not correctly received"  # Reference to HIL.dbc line 63
+        
+        time.sleep(0.06)
+        
+        # Retrieve the current power status of the brake light from the PDU HIL board
+        PowBrakeLight = pdu_hil.get_signal("PowBrakeLight") # Reference to HIL.dbc line 50
+
+        # Check if the voltage is within the range where the brake light should be off
+        if 0 <= voltage <= BRAKE_LIGHT_ON_THRESHOLD_VOLTAGE_1:
+            if PowBrakeLight:
+                assert PowBrakeLight == 0, "Brake light should be off at this voltage level"
+        
+        # Handle the unstable state range (from 833 to 841), where the brake light might be either on or off
+        elif BRAKE_LIGHT_ON_THRESHOLD_VOLTAGE_1 < voltage < BRAKE_LIGHT_ON_THRESHOLD_VOLTAGE_2:
```

Comment:
Is the reason we have this this because the output of the HIL board is noisy? Not sure I like having an if statement here. If noise is the issue, I would lower the cutoff value of the condition you have on line 44 and remove this check.

set PDU_HIL_DAC_NOISE to an appropriate number

elif voltage >= BRAKE_LIGHT_ON_THRESHOLD_VOLTAGE_2 + PDU_HIL_DAC_NOISE:

Reply:
Remember that the point of the HIL is to validate the core functionality of the car. Even if we have the threshold at 15% in the firmware, I think its acceptable if the HIL test outputs what it thinks is 20% for the test if we are battling noise. Ideally in the next revision of the board we spec a better DAC but this solution makes the most of what we currently have